presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java

```diff
@@ -90,6 +98,56 @@
     private final ServiceSelectorManager serviceSelectorManager;
     private final Announcer announcer;
     private final QueryManager queryManager;
+    private final GracefulShutdownHandler gracefulShutdownHandler;
+    private final ShutdownAction testShutdownAction;
+    private final boolean coordinator;
+
+    public static class TestShutdownAction implements ShutdownAction
+    {
+        private Lock lock = new ReentrantLock();
```

Comment:
I would replace the lock and condition with a `CountDownLatch` (initialize it to 1). Then in `onShutdown` call `countDown()` and in `waitForShutdownComplete` just wait on the latch